### PR TITLE
#560 Refactor dataset deaccession service

### PR DESCRIPTION
**Pull request by wtr-icm**

### CODE REVIEW

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionServiceTest.java [hunk 1]
```diff
@@ -18,25 +15,27 @@
...
 
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
 @MockitoSettings(strictness = Strictness.LENIENT)
 public class DatasetDeaccessionServiceTest {
```
Comment (wtr-icm) [inline dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionServiceTest.java]: I don't know if we really need this test now – I was able to refactor it easily, because of the architecture of the service, but this is only a coincidence.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    private Name extractName(Annotation[] annotations) {
+        return Arrays.stream(annotations)
+                .filter(a -> a instanceof PermissionNeeded)
+                .map(PermissionNeeded.class::cast)
+                .map(PermissionNeeded::value)
+                .map(Name::new)
+                .findFirst()
+                .orElse(Name.EMPTY);
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java]: Why we are using special indication for what we consider wrong configuration for required permission annotations?
And why we silently omit this?
Doesn't this should just throw error?

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        Map<String, PermissionNeeded> permissionMap = createPermissionMap(extractAnnotations(method));
+
+        BiConsumer<Set<RestrictedObject>, Tuple2<String, DvObject>> updateResultWithParameter = (result, param) -> {
+                PermissionNeeded config = permissionMap.computeIfAbsent(param._1,
+                        s -> { throw new IllegalStateException("No permission data for name: " + s); });
+                DvObject object = param._2;
+                result.add(createRestrictedObject(object, config));
+                if (config.needsOnOwner().length > 0) {
+                    result.add(createRestrictedOwnerObject(object, config));
+                }
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java]: imo this is more then collecting data. I would really try to include this code into `.map` or `.flatMap` in stream below

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    public Set<RestrictedObject> gatherPermissionRequirements(Method method, Object[] parameterValues) {
+        if (method == null || parameterValues == null || parameterValues.length == 0) {
+            return Collections.emptySet();
+        }
+
+        Map<String, PermissionNeeded> permissionMap = createPermissionMap(extractAnnotations(method));
+
+        BiConsumer<Set<RestrictedObject>, Tuple2<String, DvObject>> updateResultWithParameter = (result, param) -> {
+                PermissionNeeded config = permissionMap.computeIfAbsent(param._1,
+                        s -> { throw new IllegalStateException("No permission data for name: " + s); });
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java]: Your intension here is not `.computeIfAbsent`, but to throw exception if there is no key. That's why I think we should use
```
if (!...containsKey(..)) {
  throw new ...
}
```

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                null, firstMissing.required, firstMissing.dvObject);
+    }
+
+    // -------------------- PRIVATE ---------------------
+
+    private ActionLogRecord fetchOrCreateLogRecord(InvocationContext ctx, Method method, DataverseRequest request) {
+        return (ActionLogRecord) ctx.getContextData().computeIfAbsent(InterceptorCommons.LOG_RECORD_KEY,
+                k -> new ActionLogRecord(ActionType.Command, createName(method))
+                        .setUserIdentifier(request.getUser().getIdentifier()));
+    }
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java]: Creating here ActionLogRecord is tight binding of functionalities that could be completely separated. IMO logging interceptor should handle this

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        } catch (RuntimeException re) {
+            logRecord.setActionResult(Result.InternalError);
+            addExceptionMessage(logRecord, re::getMessage);
+            logConstraintViolations(logRecord, re);
+            throw re;
+        } finally {
+            if (logRecord.getActionResult() == null) {
+                logRecord.setActionResult(Result.OK);
+            } else {
+                ejbContext.setRollbackOnly();
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java:69]: Not sure if setting .setRollbackOnly is needed does ejb won't rollback when some exception will occur?

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        MissingPermissions firstMissing = missingPermissions.get(0);
+        return new PermissionException("Can't execute method " + method
+                + ", because request: " + request
+                + " is missing permissions: " + firstMissing.missing
+                + " on Object: [" + firstMissing.dvObjectName
+                + "]. All missing permissions: " + missingPermissions,
+                null, firstMissing.required, firstMissing.dvObject);
+    }
+
+    // -------------------- PRIVATE ---------------------
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java]: section comment is in wrong place. Method above is private

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionBean.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import edu.harvard.iq.dataverse.persistence.dataset.DatasetVersion.VersionState;
+import edu.harvard.iq.dataverse.search.IndexServiceBean;
+
+import javax.ejb.Stateless;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+
+@Stateless
+public class DatasetDeaccessionBean {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionBean.java]: TBH I think that functionality enclosed in this class should be moved to DatasetDeaccessionService

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/PermissionNeeded.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * <br>
+ * The usage is following: we place the annotation inside @Restricted annotation (possibly multiple times) to configure
+ * needed permissions. If there are more that one annotation on this level, attribute <b>on</b> should be used to
+ * differentiate between target parameters – and in this case corresponding parameters should have the same <b>value</b>
+ * on PARAMETER level annotation. If there is no <b>on</b> attribute filled it will be used for every parameter without
+ * filled <b>value</b>.
+ */
+@Target({ElementType.METHOD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface PermissionNeeded {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/PermissionNeeded.java:31]: Maybe add one or two examples of method signature (could be taken from tests). Something like:
```
Usage example:
<code>
        @Restricted({
                @PermissionNeeded(on = "source", needs = {Permission.EditDataset}),
                @PermissionNeeded(needs = {Permission.AddDataset})
        })
        public void method(@PermissionNeeded("source") DvObject o1, @PermissionNeeded DvObject o2)
</code>
```

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/Restricted.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+import javax.interceptor.InterceptorBinding;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@InterceptorBinding
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Restricted {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/Restricted.java:21]: Add some comment with `see {@link PermissionNeeded} for usage description`
And pointer to class handling this annotation

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionServiceTest.java (hunk 1 above)
Comment (madryk) [reply dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionServiceTest.java]: We could leave it. IMO this test doesn't really check everything that it should (missing verify of the actual arguments - now it only verifies that something was executed but it doesn't check what was the arguments).

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionBean.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+    @PersistenceContext
+    EntityManager em;
+
+    IndexServiceBean indexService;
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionBean.java]: private fields?

**Review by madryk — CHANGES_REQUESTED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java (hunk 2 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java]: Maybe method & object name is a bit misleading in this case. The point here is to match parameter annotations (which are got by reflection on the method) with parameter values (which are provided by the interception context): in case there is no PermissionNeeded annotation on param, it's simply a parameter that is not a subject of access restriction.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionBean.java (hunk 8 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionBean.java]: Good point – I put it into this class in order to be able to access it as a container-proxied method, but now I see that in this case it's not really needed.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java (hunk 5 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/RestrictedInterceptor.java]: Yes. That was a kind of defensive behavior in case of LoggedCallInterceptor issues, but in fact this would not be persisted at the end, so I can remove it completely.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java (hunk 2 above)
Comment (madryk) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/annotations/processors/PermissionDataProcessor.java]: Ok. Thanks for clarification. I've read code the wrong way. `Name.EMPTY` strongly suggest that there is an annotation, but its value is empty.
I would consider either renaming `Name.Empty` to something that will suggest that it does not exists or make signature of method return `Optional<Name>` and not use constant at all

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java (hunk 6 above)
Comment (wtr-icm) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java:69]: I've checked, and the answer is that theoretically it should work that way with some exceptions in case of caught exceptions – however because of possible corner cases I'd prefer to leave it as it is.

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionServiceTest.java (hunk 1 above)
Comment (wtr-icm) [reply dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/dataset/deaccession/DatasetDeaccessionServiceTest.java]: After restructuring DeaccessionService I found that it is not possible to retain this unit test without making completely new set of tests with different way of testing, so I've decided to remove it (we have integration tests so I think it's sufficient).

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java (hunk 6 above)
Comment (madryk) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java:69]: Ok.

**Review by madryk — APPROVED** [review]

**Review by madryk — APPROVED** [review]
